### PR TITLE
feat: switch from Haiku to Equinox

### PR DESCRIPTION
**Pull request by panahiparham**
In this work in progress PR I am working on switching from Haiku to Equinox. The main changes are,

1. There is no longer an `AgentState` attribute, instead there are `network` and `opt_state` attributes (DQN also has a `target_network`)
2. Now we use equinox and its corresponding filtered jax operations to handle neural nets creation and training

What we are currently missing,
1. Anything except ReluNets (so mainly MinatarNet and AtariNet)
2. Speed (Code runs a little slow compared to Haiku
3. Type annotations for additions to `DQN` and `EQRC`

What I have verified so far,
1. Run DQN and EQRC and MC and checked it reached good performance in 1 seed
2. Checked that action values do not explode and do not turn Nan

Please let me know about the above decisions and if you think we should handle them in a different way. Please also point out areas of improvement for speed, code quality, and bugs.

### CODE REVIEW

File: src/algorithms/nn/DQN.py [hunk 1]
```diff
@@ -82,33 +68,24 @@ def update(self):
             self.collector.collect(k, np.mean(v).item())
 
         if self.updates % self.target_refresh == 0:
-            self.state.target_params = self.state.params
+            self.target_network = copy(self.network)
```
Comment (panahiparham) [inline src/algorithms/nn/DQN.py]: Is this an appropriate way to handle target nets?

File: src/algorithms/nn/DQN.py [hunk 2]
```diff
@@ -1,26 +1,18 @@
...
+import equinox as eqx
 import chex
-import optax
 import numpy as np
-import haiku as hk
 import jax.numpy as jnp
-import utils.chex as cxu
-
-@cxu.dataclass
-class AgentState:
```
Comment (panahiparham) [inline src/algorithms/nn/DQN.py:19]: What do you think about removing the AgentState? We can add it back in and store the model params alongside optimizer parameters. To perform forward passes we would use `eqx.partition` and `eqx.combine` a lot of times. Will it be slow?

File: src/algorithms/nn/EQRC.py [hunk 3]
```diff
@@ -65,38 +57,28 @@ def update(self):
...
+    def _computeUpdate(self, network, optim, batch: Batch):
+        grad, metrics = eqx.filter_grad(self._loss, has_aux=True)(network, batch)
+        updates, new_optim = self.optimizer.update(grad, optim, eqx.filter(network, eqx.is_array))
+        params = eqx.filter(network, eqx.is_array)
 
-        updates, new_optim = self.optimizer.update(grad, state.optim, params)
-        assert isinstance(updates, dict)
-
-        decay = tree_map(
+        updates.heads['h'] = jax.tree.map(
```
Comment (panahiparham) [inline src/algorithms/nn/EQRC.py]: This step should be verified carefully.

File: src/representations/networks.py [hunk 4]
```diff
@@ -1,90 +1,61 @@
...
+
+def buildFeatureNetwork(inputs: tuple, config: dict[str, Any], key: jax.Array):
+    name = config['type']
+    hidden = config['hidden']
+
+    if name == 'TwoLayerRelu':
+        assert len(inputs) == 1
+        key_1, key_2 = jax.random.split(key, 2)
+
+        return eqx.nn.Sequential([
```
Comment (panahiparham) [inline src/representations/networks.py]: Alternatively, each of these can be its own class in `utils/eqx.py` instead of being a `Sequential`.

File: src/utils/eqx.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import jax
+import equinox as eqx
+
+class MultiHead(eqx.Module):
```
Comment (panahiparham) [inline src/utils/eqx.py]: What do you think about this? I wanted a multihead network to accomodate both QRC and DQN.

File: src/algorithms/nn/DQN.py [hunk 6]
```diff
@@ -32,29 +24,23 @@ def q_loss(q, a, r, gamma, qp):
...
         super().__init__(observations, actions, params, collector, seed)
         # set up the target network parameters
         self.target_refresh = params['target_refresh']
-
-        self.state = AgentState(
-            params=self.state.params,
-            target_params=self.state.params,
-            optim=self.state.optim,
-        )
+        self.target_network = copy(self.network)
```
Comment (artemis79) [inline src/algorithms/nn/DQN.py]: Maybe this should be deep copy instead of copy, so it copies the values of the object instead of the reference.
